// turbo-cli/dist/default/vue/webpack.config.js
const webpack = require('webpack')
const path = require('path')
const UglifyJsWebpackPlugin = require('uglifyjs-webpack-plugin')
const VueLoaderPlugin = require('vue-loader/lib/plugin')

module.exports = {
    entry: {
		app: './src/index.js'
	},
    output: {
		path: __dirname+'/public/dist',
		filename: 'bundle/[name].js',
        sourceMapFilename: 'bundle/[name].map',
        chunkFilename: 'bundle/[name].js'
    },
	devtool: '#source-map',
	performance: process.env.NODE_ENV === 'production' ? {hints: false} : {},
	plugins: process.env.NODE_ENV === 'production' ? [
	    new webpack.DefinePlugin({
	        'process.env': {
	        	'NODE_ENV': JSON.stringify('production')
	        }
	    })
	] : [],

	optimization: process.env.NODE_ENV === 'production' ? {
		minimize: true,
		minimizer: [
			new UglifyJsWebpackPlugin({
				uglifyOptions: {
					output: {
						comments: false
					},
					compress: {
						drop_console: true,
						dead_code: true
					}
				}
			})
		]
	} : {
		minimize: false,
	},
	node: {
		fs: 'empty'
	},
    plugins: [new VueLoaderPlugin()],
    module: {
        rules: [
            {
                test: /\.vue$/,
                loader: "vue-loader"
            },
            {
                test: /\.(scss|css)$/,
                use: ["vue-style-loader", "css-loader", "sass-loader"]
            },
            {
                test: /\.js$/,
                loader: "babel-loader",
                exclude: /node_modules/
            }
        ]
    },
    resolve: {
        alias: {
            vue$: "vue/dist/vue.esm.js"
        },
        extensions: ["*", ".js", ".vue", ".json"]
    }
}
