// vertex360/src/index.js
var superagent = require('superagent')
var Promise = require('bluebird')

// Vertex imports
var Vertex = require('./vertex/Vertex')
var Router = require('./vertex/Router')
var local = require('./vertex/local')
// var graphQl = require('./graph-ql')

var Microservice = function(credentials){
	const BASE_URL = 'https://api.turbo360.co'
	const TURBO_URL = 'https://www.turbo360.co'

	var config = {
		site_id: credentials['site_id'],
		turbo_url: TURBO_URL, // for fetching turbo dashboard resources (site, profile, etc)
		base_url: BASE_URL
	}

	var vertexApp = function(opts){ // opts can be null
		const keys = Object.keys(process.env)
		keys.forEach(function(key, i) {
			if (key.indexOf('AWS_') != -1)
				delete process.env[key]
		})
		
		delete process.env['PATH']
		delete process.env['_HANDLER']
		delete process.env['LD_LIBRARY_PATH']
		delete process.env['LAMBDA_TASK_ROOT']
		delete process.env['LAMBDA_RUNTIME_DIR']

		if (process.env.TURBO_ENV == 'prod'){ // this is null in production
			var app = new Vertex(opts)
			return app
		}

		var app = local.app(opts)
		return app
	}

	var router = function(){
		if (process.env.TURBO_ENV == 'prod'){ // this is null in production
			var router = new Router()
			return router
		}

		// console.log('DEV ROUTER!')
		var router = local.router()
		return router
	}

	// var graphqlServer = function(){
	// 	return graphQl
	// }

	var client = {
		express: vertexApp,
		app: vertexApp,
		router: router
		// graphqlServer: graphqlServer
	}

	return client
}

module.exports = Microservice
