// turbo-cli/dist/extend/api/models/Post.js
/*
	Sample blog post schema
*/

const mongoose = require('mongoose')

const PostSchema = new mongoose.Schema({
	profile: {type:mongoose.Schema.Types.Mixed, default:{}},
	site: {type:String, trim:true, default: ''}, // id number of site
	title: {type:String, trim:true, default: ''},
	preview: {type:String, trim:true, default: ''},
	tags: {type:Array, default:[]},
	text: {type:String, trim:true, default: ''},
	image: {type:String, trim:true, default: ''},
	slug: {type:String, lowercase:true, trim:true, default:''},
	link: {type:String, trim:true, default:''},
	type: {type:String, lowercase:true, default:'original'}, // original or link
	numReplies: {type:Number, default:0},
	isPublic: {type:String, trim:true, default:'no'},
	dateString: {type:String, default:'' },
	timestamp: {type:Date, default:Date.now}
})

PostSchema.methods.summary = function() {
	const summary = {
		title: this.title,
		tags: this.tags,
		image: this.image,
		profile: this.profile,
		preview: this.preview,
		text: this.text,
		slug: this.slug,
		isPublic: this.isPublic,
		numReplies: this.numReplies,
		link: this.link,
		type: this.type,
		dateString: this.dateString,
		timestamp: this.timestamp,
		schema: 'post',
		id: this._id.toString()
	}

	return summary
}

PostSchema.statics.resourceName = 'post'

PostSchema.statics.convertToJson = function(posts){
	return posts.map(function(p){
		return p.summary()
	})
}

module.exports = mongoose.model('PostSchema', PostSchema)
